api/v1alpha1/kongplugin_installation_types.go

```diff
@@ -0,0 +1,129 @@
+//+kubebuilder:printcolumn:name="Ready",type=string,JSONPath=`.status.conditions[?(@.type=="Ready")].status`
+
+// KongPluginInstallation allows using a custom Kong Plugin distributed as a container image available in a registry.
+// Such a plugin can be associated with GatewayConfiguration or DataPlane to be available for particular Kong Gateway
+// and configured with KongPlugin CRD.
+type KongPluginInstallation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KongPluginInstallationSpec   `json:"spec,omitempty"`
+	Status KongPluginInstallationStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// KongPluginInstallationList contains a list of KongPluginInstallation.
+type KongPluginInstallationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KongPluginInstallation `json:"items"`
+}
+
+// KongPluginInstallationSpec defines the desired state of KongPluginInstallation.
+type KongPluginInstallationSpec struct {
+
+	// The image is an OCI image URL for a packaged Custom Kong Plugin.
+	//
+	//+kubebuilder:validation:Required
+	Image string `json:"image"`
+
+	// SecretRef allows referring specific Kubernetes Secret to use for OCI registry authentication for pulling an image with Custom Kong Plugin.
+	// The Secret format should follow https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry.
+	// When the field is omitted it is assumed that the image is public and can be fetched without providing credentials.
+	//
+	//+optional
+	SecretRef *corev1.SecretReference `json:"secretRef,omitempty"`
+}
+
+// KongPluginInstallationStatus defines the observed state of KongPluginInstallation.
+type KongPluginInstallationStatus struct {
+	// Conditions describe the current conditions of this KongPluginInstallation.
+	//
+	//+listType=map
+	//+listMapKey=type
+	//+kubebuilder:validation:MaxItems=8
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// KongPluginInstallationConditionType is the type for status conditions on
+// KongPluginInstallation resources. This type should be used with the
+// KongPluginInstallationStatus.Conditions field.
+type KongPluginInstallationConditionType string
+
+// KongPluginInstallationConditionReason defines the set of reasons that explain why
+// a particular KongPluginInstallation condition type has been raised.
+type KongPluginInstallationConditionReason string
+
+const (
+	// This condition indicates whether the controller has fetched the plugin image
+	// and made it available for use as a specific Custom Kong Plugin.
+	//
+	// It is a positive-polarity summary condition, and so should always be
+	// present on the resource with ObservedGeneration set.
+	//
+	// It should be set to Unknown if the controller performs updates to the
+	// status before it has all the information it needs to be able to determine
+	// if the condition is true.
```

Comment:
Retrieval and installation should probably be separate Condition types, since they can fail or succeed independently. Likely scenarios include:

- The controller cannot and has never retrieved the plugin image, e.g. because the credentials are invalid. Retrieval is `False` and installation is also `False`.
- The controller has retrieved the plugin image successfully, but cannot install it, e.g. because the API server died in the interim between retrieval and the Deployment modification attempt (ignore that the controller won't be able to update status _either_ in this case--there just are scenarios where retrieval is fine, but the work after can't proceed). Retrieval is `True` but installation is `False`.
- The controller has previously retrieved and installed the plugin, but the image server goes offline and the controller can't check it for updates, so retrieval is currently `False` while installation is `True`.

Do we have a rough implementation outline for the binding UX between DataPlanes/Gateways and KongPluginInstallations? This PR should probably include scaffolding for that alongside the CRD, since the binding semantics matter for status information. The design doc doesn't cover that AFAICT.

The main point I can think of is whether the KongPluginInstallation to DataPlane/Gateway relationship is many to one or many to many. The doc implies multiple KongPluginInstallations (it only allows for a single image in a given CR, and I expect we want to support multiple custom plugins), but doesn't indicate if they can then be requested by multiple DataPlane/Gateway resources or only one.

The many to one case is simple, as the installation type condition need only reflect a single modified Deployment, but the many to many case is less simple. If we support many to many, we'd need either:
- Installation is an aggregate condition. All requesters satisfied (`True`) or not all requesters satisfied (`False`). We'd presumably need to indicate _which_ requesters are unsatisfied in `message`, since it's the only dynamic value field.
- Installation either isn't a condition or is an aggregate condition, but `Status` includes an additional field with per-requester information.

I kinda considered a separate `InstallationConditions` array for per-instance, but there's not really any field in `Condition` that'd make sense for storing the instance identifier.

Reply:
It's like that, I've added examples to descriptions.